ComputePipeline/src/ComputePipeline.jl

```diff
@@ -0,0 +1,384 @@
+        println("    ", k, "=>", out)
+    end
+    return println(io, ")")
+end
+
+function ComputeGraph()
+    return ComputeGraph(Dict{Symbol,ComputeEdge}(), Dict{Symbol,Computed}())
+end
+
+function isdirty(computed::Computed)
+    hasparent(computed) || return false
+    parent = computed.parent
+    # Can't be dirty if inputs have changed
+
+    if parent.got_resolved[]
+        # if resolved is true, the computed value is dirty if the output is dirty
+        return computed.parent.outputs_dirty[computed.parent_idx]
+    else
+        # if resolved is false, the computed value is dirty if any of the inputs have changed
+        return any(parent.inputs_dirty)
+    end
+end
+
+function isdirty(edge::ComputeEdge)
+    # If resolve hasn't run, it has to be dirty
+    edge.got_resolved[] || return true
+    # Otherwise it's dirty if the input changed
+    return any(edge.inputs_dirty)
+end
+
+function mark_dirty!(edge::ComputeEdge)
+    edge.got_resolved[] = false
+    for dep in edge.dependents
+        mark_dirty!(dep)
+    end
+    return
+end
+
+function mark_dirty!(computed::Computed)
+    hasparent(computed) || return
+    return mark_dirty!(computed.parent)
+end
+
+function resolve!(input::Input)
+    if isassigned(input.output.value)
+        input.output.value[] = input.f(input.value)
+    else
+        input.output.value = RefValue(input.f(input.value))
+    end
+    input.dirty = false
+    return input.output.value[]
+end
+
+function mark_dirty!(input::Input)
+    input.dirty = true
+    for edge in input.dependents
+        mark_dirty!(edge)
+    end
+    return
+end
+
+function Base.setindex!(computed::Computed, value)
+    computed.value[] = value
+    return mark_dirty!(computed)
+end
+
+function Base.setproperty!(attr::ComputeGraph, key::Symbol, value)
+    input = attr.inputs[key]
+    input.value = value
+    mark_dirty!(input)
+    return value
+end
+
+Base.haskey(attr::ComputeGraph, key::Symbol) = haskey(attr.inputs, key)
+
+function Base.getproperty(attr::ComputeGraph, key::Symbol)
+    # more efficient to hardcode?
+    key === :inputs && return getfield(attr, :inputs)
+    key === :outputs && return getfield(attr, :outputs)
+    key === :default && return getfield(attr, :default)
+    return attr.inputs[key].output
+end
+
+function Base.getindex(attr::ComputeGraph, key::Symbol)
+    return attr.outputs[key]
+end
+
+Base.getindex(computed::ComputedValue) = resolve!(computed)
+
+function mark_input_dirty!(parent::ComputeEdge, edge::ComputeEdge)
+    for (i, input) in enumerate(edge.inputs)
+        # This gets called from resolve!(parent), so we should only mark dirty if the input is a child of parent
+        hasparent(input) && input.parent === parent && (edge.inputs_dirty[i] = isdirty(input))
+    end
+end
+
+function set_result!(edge::TypedEdge, result, i, value)
+    if isnothing(value)
+        edge.outputs_dirty[i] = false
+    else
+        edge.outputs_dirty[i] = true
+        edge.outputs[i][] = value
+    end
```

Comment:
If I understand correctly this interprets `nothing` as "the value has not changed". In that case shouldn't it leave the dirty state untouched? If a previous call has dirtied output i and this call has not changed it, it should still dirty, right?

```suggestion
    if !isnothing(value)
        edge.outputs_dirty[i] = true
        edge.outputs[i][] = value
    end
```

Reply:
>In that case shouldn't it leave the dirty state untouched? If a previous call has dirtied output i and this call has not changed it, it should still dirty, right?

No, the dirty state is local to the current resolve. 
So needs to be set to the correct value when setting the new value.
We still need to double check if there's a simpler and more consistent way and add some tests if this works 100% as intended, but right now `outputs_dirty` is only for the children of an edge to figure out _after_ a resolve if a value has changed from that particular resolve.